### PR TITLE
Allow CPU self.min_val to be moved to GPU

### PR DESCRIPTION
**Pull request by junesg**
Summary: Right after seeding a fake quantization model, we noticed that self.min_val and self.max_val are loaded from seed model into current model on CPU. During run time, errors showed up because the self.min_val and self.max_val of fake quantization is right now assumed to be initialized on GPU during first run-step.

Differential Revision: D30767803

### CODE REVIEW

File: torch/quantization/observer.py [hunk 1]
```diff
@@ -532,8 +538,12 @@ def forward(self, x_orig):
             min_val, max_val = torch._aminmax(x)
         else:
             min_val_cur, max_val_cur = torch._aminmax(x)
-            min_val = min_val + self.averaging_constant * (min_val_cur - min_val)
-            max_val = max_val + self.averaging_constant * (max_val_cur - max_val)
+            min_val = min_val.to(device=min_val_cur.device) + \
```
Comment (supriyar) [inline torch/quantization/observer.py:541]: Can the code to move `min_val`/`max_val` be moved outside this statement so we do it only once?

**Review by supriyar — COMMENTED** [review]
Is there a small example/gist that can show the bug? We could add this as a unit test to `test/quantization/core/test_workflow_module.py`

**Comment by vkuzo** [comment]
@junesg thanks for finding the issue, this has been something people complained about for months!

As far as the fix, did I understand you correctly that the problem occurs when loading from state_dict, min_val/max_val/eps are not respecting device?  If yes, could we look into fixing `FakeQuantize._load_from_state_dict` and `ObserverBase._load_from_state_dict` functions instead?

**Comment by junesg** [comment]
> @junesg thanks for finding the issue, this has been something people complained about for months!
> 
> As far as the fix, did I understand you correctly that the problem occurs when loading from state_dict, min_val/max_val/eps are not respecting device? If yes, could we look into fixing `FakeQuantize._load_from_state_dict` and `ObserverBase._load_from_state_dict` functions instead?

Thank you @vkuzo , this is a very good suggestion! What is not immediately clear to me is how the self.min_val/max_val/eps/scales are distributed to the different GPUs during distributed training right after state loading. When we load the model, we loaded it to cpu -- are you suggesting that in `FakeQuantize._load_from_state_dict` we manually ensure the self.values are propagated into all devices?

**Comment by vkuzo** [comment]
Loading the state dict is expected to put things on the correct device: https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-model-across-devices .  This should happen somewhere in `nn.Module::_load_from_state_dict`.

Do you have a way to reproduce the issue with a toy model?

**Comment by junesg** [comment]
Let me try to set up a toy model -- the tricky part is to try on cpu + 2 GPU for the toy model. 
I will also try state loading to GPU. will loop back.

File: test/quantization/core/test_workflow_module.py [hunk 2]
```diff
@@ -899,6 +900,57 @@ def forward(self, x):
         input = torch.randn(4, 1, 4, 4, device=device)
         model(input)
 
+    @unittest.skipIf(not TEST_MULTIGPU, "multi-GPU not supported")
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    def test_model_parallel(self):
```
Comment (vkuzo) [inline test/quantization/core/test_workflow_module.py:905]: thanks for creating the test case! To confirm, does this test case fail before the fix and pass after the fix?

File: test/quantization/core/test_workflow_module.py [hunk 3]
```diff
@@ -899,6 +900,57 @@ def forward(self, x):
...
+        with override_quantized_engine('fbgemm'):
+            device = torch.device('cpu')
+            model_to_save = self.dummyModel(device)
+
+            temp_file_name = os.path.join(
+                tempfile._get_default_tempdir(), "temp_stat_dict.pt"
+            )
+            torch.save(model_to_save.state_dict(), temp_file_name)
+            temp_states = torch.load(temp_file_name,  map_location=(lambda s, _: torch.serialization.default_restore_location(s, 'cpu')))
+            model = torch.device('gpu')
```
Comment (vkuzo) [inline test/quantization/core/test_workflow_module.py:938]: is this intended?

**Comment by junesg** [comment]
After writing and polishing unit test offline. I fixed this bug at a higher level. Note that this is NOT a pytorch bug, rather the user or wrappers on top of pytorch is responsible to distribute model to different GPU devices.
